runtime/ast/access.go

```diff
@@ -20,38 +20,167 @@ package ast
 
 import (
 	"encoding/json"
+	"strings"
 
 	"github.com/onflow/cadence/runtime/errors"
 )
 
-//go:generate go run golang.org/x/tools/cmd/stringer -type=Access
+//go:generate go run golang.org/x/tools/cmd/stringer -type=PrimitiveAccess
 
-type Access uint
+type Access interface {
+	isAccess()
+	Keyword() string
+	Description() string
+	String() string
+	MarshalJSON() ([]byte, error)
+	IsLessPermissiveThan(Access) bool
+}
+
+type EntitlementSet interface {
+	Entitlements() []*NominalType
+	Separator() string
+}
+
+type ConjunctiveEntitlementSet struct {
+	Elements []*NominalType `json:"ConjunctiveElements"`
+}
+
+var _ EntitlementSet = ConjunctiveEntitlementSet{}
+
+func (s ConjunctiveEntitlementSet) Entitlements() []*NominalType {
+	return s.Elements
+}
+
+func (s ConjunctiveEntitlementSet) Separator() string {
+	return ","
+}
+
+func NewConjunctiveEntitlementSet(entitlements []*NominalType) ConjunctiveEntitlementSet {
+	return ConjunctiveEntitlementSet{Elements: entitlements}
+}
+
+type DisjunctiveEntitlementSet struct {
+	Elements []*NominalType `json:"DisjunctiveElements"`
+}
+
+var _ EntitlementSet = DisjunctiveEntitlementSet{}
+
+func (s DisjunctiveEntitlementSet) Entitlements() []*NominalType {
+	return s.Elements
+}
+
+func (s DisjunctiveEntitlementSet) Separator() string {
+	return " |"
+}
+
+func NewDisjunctiveEntitlementSet(entitlements []*NominalType) DisjunctiveEntitlementSet {
+	return DisjunctiveEntitlementSet{Elements: entitlements}
+}
+
+type EntitlementAccess struct {
+	EntitlementSet EntitlementSet
+}
+
+var _ Access = EntitlementAccess{}
+
+func NewEntitlementAccess(entitlements EntitlementSet) EntitlementAccess {
+	return EntitlementAccess{EntitlementSet: entitlements}
+}
+
+func (EntitlementAccess) isAccess() {}
+
+func (EntitlementAccess) Description() string {
+	return "entitled access"
+}
+
+func (e EntitlementAccess) entitlementsString(prefix strings.Builder) strings.Builder {
```

Comment:
Maybe just pass by pointer here and avoid the return type, and having to remember to use the result at the call site
```suggestion
func (e EntitlementAccess) entitlementsString(prefix *strings.Builder) {
```